scanpy/tools/_rank_genes_groups.py

```diff
@@ -177,275 +387,84 @@ def rank_genes_groups(
-                scores[np.isnan(scores)] = 0
-                pvals = 2 * stats.distributions.norm.sf(np.abs(scores))
-
-                if corr_method == 'benjamini-hochberg':
-                    pvals[np.isnan(pvals)] = 1  # set Nan values to 1 to properly convert using Benhjamini Hochberg
-                    _, pvals_adj, _, _ = multipletests(pvals, alpha=0.05, method='fdr_bh')
-                elif corr_method == 'bonferroni':
-                    pvals_adj = np.minimum(pvals * n_genes, 1.0)
-
-                # Fold change
-                foldchanges = (expm1_func(means[imask]) + 1e-9) / (expm1_func(mean_rest) + 1e-9)  # add small value to remove 0's
-                scores_sort = np.abs(scores) if rankby_abs else scores
-                partition = np.argpartition(scores_sort, -n_genes_user)[-n_genes_user:]
-                partial_indices = np.argsort(scores_sort[partition])[::-1]
-                global_indices = reference_indices[partition][partial_indices]
-                rankings_gene_scores.append(scores[global_indices])
-                rankings_gene_names.append(adata_comp.var_names[global_indices])
-                rankings_gene_logfoldchanges.append(np.log2(foldchanges[global_indices]))
-                rankings_gene_pvals.append(pvals[global_indices])
-                rankings_gene_pvals_adj.append(pvals_adj[global_indices])
-
-        # If no reference group exists,
-        # ranking needs only to be done once (full mask)
-        else:
-            scores = np.zeros((n_groups, n_genes))
-            chunk = []
-            n_cells = X.shape[0]
-            n_genes_max_chunk = floor(CONST_MAX_SIZE / n_cells)
-            if n_genes_max_chunk < n_genes:
-                chunk_index = n_genes_max_chunk
-                while chunk_index < n_genes:
-                    chunk.append(chunk_index)
-                    chunk_index = chunk_index + n_genes_max_chunk
-                chunk.append(n_genes)
-            else:
-                chunk.append(n_genes)
-            left = 0
-            for chunk_index, right in enumerate(chunk):
-                # Check if issparse is true
-                if issparse(X):
-                    df1 = pd.DataFrame(data=X[:, left:right].todense())
-                else:
-                    df1 = pd.DataFrame(data=X[:, left:right])
-                ranks = df1.rank()
-                # sum up adjusted_ranks to calculate W_m,n
-                for imask, mask in enumerate(groups_masks):
-                    scores[imask, left:right] = np.sum(ranks.loc[mask, :])
-                left = right
+        if mean_group is not None and mean_rest is not None:
+            foldchanges = (expm1_func(mean_group) + 1e-9) / (
+                expm1_func(mean_rest) + 1e-9
+            )  # add small value to remove 0's
+            if 'logfoldchanges' not in d:
+                d['logfoldchanges'] = {}
+            d['logfoldchanges'][group_name] = np.log2(foldchanges[global_indices])
```

Comment:
Sweet. Does this also produce logfoldchanges for logreg? sc.get.rank_genes_groups_df throws an exception now:

![image](https://user-images.githubusercontent.com/1140359/76791440-4dba5080-6797-11ea-90c4-0a8da8e93102.png)

Reply:
Thank you for catching this. I'll check why it happens.

Reply:
Oh sorry for misunderstanding. Error I posted is from stable scanpy, not from the PR. I was aşkı g if it's fixed now.

Reply:
This is the problem with `sc.get`. logreg only has  `names` and `scores` keys, but `sc.get` tries to query nonexistent  'logfoldchanges' and so on.

Reply:
Should i add `logfoldchanges` for logreg?

Reply:
I don't want to interrupt the PR, but I think logfoldchange and percentages should be available for every method.

Reply:
Should the percentage for the reference group be also calculated and stored?

Reply:
I think so. It might look unnecessary for one-vs-rest kind of tests but for clusterA-vs-clusterB kind of tests it'd be useful.